### PR TITLE
Up to v24

### PR DESCRIPTION
**Pull request by Jakllp**
New:
- Accounted for higher velocity in air compared to land
- Adjusted climbing-logic for slabs
- /reloaducars now also reloads the lang-file
- You can now put "-<JumpAmount>" behind a Jumpblock in the config to overwrite the global jump amount for said block

Fixes:
- /reloaducars now actually reloads the config
- Better Spawning of cars (right on the block - offset is no more)
- Hovering on slabs reduced
- Fixed stair climbing for low speeds

Tweaks:
- Some height adjustments for climbing

### CODE REVIEW

File: src/main/java/com/useful/ucars/uCarsListener.java
Comment (storm345) [inline src/main/java/com/useful/ucars/uCarsListener.java:1361]: Why are you reducing xz speed to 65% when y is not exactly equal to integer? (Which depending on the precision used to compare double (getY()) and int (getBlockY()) may never be true. I don't see why being on a block slightly elevated in y (eg. a slab) should arbitrarily slow down the car

File: src/main/java/com/useful/ucars/uCarsListener.java
Comment (storm345) [inline src/main/java/com/useful/ucars/uCarsListener.java:1373]: Same here

File: src/main/java/com/useful/ucars/uCarsListener.java
Comment (storm345) [inline src/main/java/com/useful/ucars/uCarsListener.java:2181]: Rather than duplicating code (Bad practice) can we instead change it so that this code is in one place and one place only (Perhaps an init() method) and called where needed instead? (I know this plugin already has bits of duplicated code, but it's not a good pattern to continue)

**Review by storm345 — CHANGES_REQUESTED** [review]
See comments, I don't understand the velocity scaling thing and a point on code quality (Repetition)

File: src/main/java/com/useful/ucars/uCarsListener.java
Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:1361]: I do this to reduce/counter the effect of losing friction. When you were driving on slaps/climbing stuff/driving own something you were going quite a bit faster than when just driving on ground. Reducing the speed to 65% fixes this

File: src/main/java/com/useful/ucars/uCarsListener.java
Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:2181]: Yeah I hated to do that but for some reason I couldn't think of something like init()... I really don't know why

File: src/main/java/com/useful/ucars/uCarsListener.java
Comment (storm345) [reply src/main/java/com/useful/ucars/uCarsListener.java:1254]: Hmm... I think this is an error and that this is just reducing speed of cars perhaps even everywhere for no good reason (The config should be used for that). If you are seeing a weird speed change it'll be due to a bug somewhere else (Perhaps multiplying xyz when meaning to only multiply xz components for example). I would wager this code is probably getting called even when driving on flat ground since 0.000000001 != 0

Comment (storm345) [reply src/main/java/com/useful/ucars/uCarsListener.java:1254]: If there is such a bug, seek to fix it rather than adding "bandaid" code to address only the symptoms

Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:1254]: I've done some testing and... I do not believe that this is an issue with ucars.

I let Minecraft move the Minecart with a vector of (1,0,0) - expectation: Just going into the X direction - Result: moving slightly downwards (seemingly y: -0.04) when moving through air.
I setup a "test-track" (straight line) and measured the time it took me to pass it. Slightly over 7 seconds.
I then tested (1,1,0) and (1,0.04,1) (the latter let's the minecart "glide" - no downward movement).
Both vectors resulted in a time of around 3.5 seconds on the track.

This to me looks like some sort of "friction", friction being pressure onto the ground, is being applied by Minecraft which is obviously missing when there is no "pressure" on the minecart.

Or am I misinterpreting something here.

Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:1254]: I've done some testing and... I do not believe that this is an issue with ucars.

I let Minecraft move the Minecart with a vector of (1,0,0) - expectation: Just going into the X direction - Result: moving slightly downwards (seemingly y: -0.04) when moving through air.
This behavior can also be observed when a Minecart is frozen (there was actually an issue with this which I found during testing but just setting the vector to 0 doesn't freeze you whilst (0,0.04,0) does)

I setup a "test-track" (straight line) and measured the time it took me to pass it. Slightly over 7 seconds.
I then tested (1,1,0) and (1,0.04,1) (the latter let's the minecart "glide" - no downward movement).
Both vectors resulted in a time of around 3.5 seconds on the track.

This to me looks like some sort of "friction", friction being pressure onto the ground, is being applied by Minecraft which is obviously missing when there is no "pressure" on the minecart.

Or am I misinterpreting something here.

(Tested on 1.16.5)

Comment (storm345) [reply src/main/java/com/useful/ucars/uCarsListener.java:1254]: Was this testing with ucars not installed?

Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:1254]: I tested it with and without uCars being installed, which is why I'm so confident that it's not an uCars error.

The adjustment only happens whilst climbing/going down slopes or whilst driving on non-whole blocks.
When driving on whole blocks car.getLocation().getY() and getBlockY() do actually return the same value.

One note:
I have another commit ready but I'm holding off of commiting it as long as this isn't resolved as that leaves me the option to change something if necessary without creating unnecessary commits

Comment (storm345) [reply src/main/java/com/useful/ucars/uCarsListener.java:1254]: Okay interesting. What about when driving on carpet? Or slabs? In those cases blockY is not the same as locationY? Should the "adjustment" (Reducing 'speed' requested from the game by 45%) be applied there or not? (I think it may still also be prudent to compare "Math.abs(blockY-locationY) < eps" instead of checking double and integer equality which is usually frowned upon as an additional tweak).

Comment (storm345) [reply src/main/java/com/useful/ucars/uCarsListener.java:1254]: Maybe also add a comment about why this is needed into the code (And move the "Move the car and adjust vector..." comment to be back next to the code it is relevant to)

Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:1254]: Will add the comments.

The speed-increase also happens whilst driving on carpet/slabs as we are kind of floating on them - the blockY-Location is always an integer.

But I have to admit that I don't know what you mean with eps - could you explain?

Comment (storm345) [reply src/main/java/com/useful/ucars/uCarsListener.java:1254]: Maybe a more robust check would be to check if we are asking for y velocity component > 0 or >= 0.04 in the velocity calculated by ucars and in that case we know the 'friction' is not applied. And the reverse aswell naturally.

And I know blockY location is always an integer, that's the problem. When you compare double and int the int will be promoted to a double for the comparison and it is generally not recommended to ever compare doubles for equality. By "eps" I meant "small threshold" (eps often refers to greek letter epsilon which is often used to denote small quantities in maths). The reason is that computers do not store floating point (even double precision floating point) numbers to infinite precision and 1.999999999 != 2.000000001
https://howtodoinjava.com/java-examples/correctly-compare-float-double/

**Comment by Jakllp** [comment]
After exploring the idea you had and some more thinking I think I came up with a pretty good solution.

We now climb by looking at the height of the block ahead (the +0.2 makes sure we actually get on top of it as you would sometimes get stuck mid air)
This get's rid of checking for slabs or carpets or basically anything.
We also avoid "floating" as we won't try to climb a block that is the same height as the current block.
This results in normal friction being applied on such blocks.

So what about climbing and going down slopes.
If we climb something calculated is set to true -> easy check
If we fall down somewhere there is fallDistance added to the car. So we check for that.

I also had to change from 0.65 to 0.55 after some more testing of the actual speed difference.

**Comment by storm345** [comment]
> After exploring the idea you had and some more thinking I think I came up with a pretty good solution.
> 
> We now climb by looking at the height of the block ahead (the +0.2 makes sure we actually get on top of it as you would sometimes get stuck mid air)
> This get's rid of checking for slabs or carpets or basically anything.
> We also avoid "floating" as we won't try to climb a block that is the same height as the current block.
> This results in normal friction being applied on such blocks.
> 
> So what about climbing and going down slopes.
> If we climb something calculated is set to true -> easy check
> If we fall down somewhere there is fallDistance added to the car. So we check for that.
> 
> I also had to change from 0.65 to 0.55 after some more testing of the actual speed difference.

Nice! Using the bounding box is a good use of newer API - I like that change a lot. Looking on the apidocs for that method it says it is approximate and gives stairs as an edge case - can you check that the vehicle is still able to correctly handle stairs? (Hopefully the maxY() from the bounding box will be fine but let's check to be sure).

It is a little odd and annoying that we need to be adjusting for Minecraft velocity tweaks, but if that is what is required I guess it is alright. If we're committing to doing this can we just move the 0.04 and 0.55 magic numbers into labelled constants? Eg. GRAVITY_Y_VELOCITY_MAGNITUDE, and SIMULATED_FRICTION_SPEED_MULTIPLIER or similar? (And any others you have) It will make the code a little cleaner and easier to change with future Minecraft changes instead of magic repeated unlabelled numbers 😛

**Comment by storm345** [comment]
(closing was a misclick haha)

**Comment by Jakllp** [comment]
Checked stairs, speed is right, climbing works well

**Comment by Jakllp** [comment]
Just a sidenote: Minecraft server versions handle Minecarts differently than single player and in general Minecarts are strange.
This for example is a "bug" I found:
https://bugs.mojang.com/browse/MC-205117

**Comment by storm345** [comment]
Hmmm if that is the case, could you please add a check if the entity being used as the car is a Minecart before applying the speed adjustments? It's possible to use uCars with different entities (minecarts are just the default). For example with uCarsTrade you can instead use armor stands - which I assume don't have the same velocity weirdness?

**Comment by Jakllp** [comment]
I tested it with armor-stands and the strange slow down did not apply so I'm guessing it's really just Minecarts

**Comment by storm345** [comment]
It's looking good! Can you check the y velocity tweak (0.04 instead of 0) aswell? I suspect that needs to be 0 for armor stands. After that I think it's good to merge!

**Comment by Jakllp** [comment]
Yeah forgot about that...

File: src/main/java/com/useful/ucars/uCarsCommandExecutor.java
Comment (aimiebell) [inline src/main/java/com/useful/ucars/uCarsCommandExecutor.java:25]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Class names should start with an upper case letter (NM_CLASS_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-class-names-should-start-with-an-upper-case-letter-nm-class-naming-convention).
Class names should be nouns, in mixed case with the first letter of each internal word capitalized. Try to keep your class names simple and descriptive. Use whole words-avoid acronyms and abbreviations (unless the abbreviation is much more widely used than the long form, such as URL or HTML).

**Pull request merged**
